Implement run for system containers

giuseppe opened the pull request:
## Description

Allows to run commands in a container.  This is more useful for system containers as it allows to run a command before the container is running.  To do so, a new `config.json` file is created with a modified `args` block.

## Related Issue Numbers

https://github.com/projectatomic/atomic-system-containers/pull/89

## Pull Request Checklist:

If your Pull request contains new features or functions, tests are required. If the PR is a bug fix and no tests exist, please consider adding some to prevent regressions.
- [x] Unittests
- [x] Integration Tests

giuseppe commented:
@ashcrow @TomasTomecek

rhatdan commented:
Why do you need this?  And how is it different then atomic run?  Atomic run execs commands inside of running containers?  I think this further confuses people.  atomic is supposed to have a different CLI then docker.

giuseppe commented:
@rhatdan this is not a replacement for `atomic run`.  `atomic run` runs the RUN label of an image, where `atomic exec` is to launch an arbitrary command inside of a container (exactly what `docker exec` and `runc exec` do).

There is a difference in the system container case, as we allow to `exec` also when we don't have a running container.  We do this with a new `config.json` file that execs the specified command.

The main use case of this is https://github.com/projectatomic/atomic-system-containers/pull/89

We need to execute some commands between the installation and the service startup.  Unfortunately net-snmp doesn't allow its configuration while the daemon is running, so we need to exec into the container before we can use `runc exec`.

rhatdan commented:
Lets have conversation on  projectatomic/atomic-system-containers#89

giuseppe commented:
bot, please retry

ashcrow commented:
I'll have some unittests to add ot this today to help move this forward.

ashcrow commented:
```
+ atomic --debug run test-system-container-26530
args must not be empty

Traceback (most recent call last):
  File "/usr/bin/atomic", line 185, in <module>
    sys.exit(_func())
  File "/usr/lib/python2.7/site-packages/Atomic/run.py", line 90, in run
    return be.run(con_obj, atomic=self, args=self.args)
  File "/usr/lib/python2.7/site-packages/Atomic/backends/_ostree.py", line 169, in run
    return self.syscontainers.container_exec(name, args.detach, args.command)
  File "/usr/lib/python2.7/site-packages/Atomic/syscontainers.py", line 1973, in container_exec
    return util.check_call(cmd, stdin=sys.stdin, stderr=sys.stderr, stdout=sys.stdout)
  File "/usr/lib/python2.7/site-packages/Atomic/util.py", line 197, in check_call
    return subprocess.check_call(cmd, env=env, stdin=stdin, stderr=stderr, stdout=stdout, close_fds=True)
  File "/usr/lib64/python2.7/subprocess.py", line 542, in check_call
    raise CalledProcessError(retcode, cmd)
CalledProcessError: Command '['/bin/runc', 'run', 'test-system-container-26530']' returned non-zero exit status 1
+ teardown
```

giuseppe commented:
@ashcrow I've changed it to reuse `atomic run` now instead of adding a new command to atomic.

I've added a fixup patch to address that error

ashcrow commented:
@giuseppe 1365d6523d738823a9d3f986c9d741761df626f5

giuseppe commented:
@ashcrow thanks!  I've cherry picked the patch

ashcrow commented:
@giuseppe looks like the fixup changes one of the unittests. Will add a follow on to update it.

ashcrow commented:
How odd. I can't replicate the failure:

```
TEST_UNIT=syscontainers ./test.sh
UNIT TESTS:                                                                                               
...tests/unit/test_syscontainers.py:100: DeprecationWarning: Please use assertEqual instead.              
  self.assertEquals(_cc.call_args, expected_call)                                                         
..                                                                                                        
----------------------------------------------------------------------                                    
Ran 5 tests in 0.038s                                                                                     
                                                          
```

ashcrow commented:
bot, retest this please

ashcrow commented:
I tried running the trust test cases as well and they worked (well, some paths were wrong but that's because I'm running in a virtualenv as non-root).

jlebon commented:
The CentOS one died due to infra flake. I restarted just that one.

ashcrow commented:
@jlebon thanks!

ashcrow commented:
@giuseppe ad7ab9b6178f327ea94af15fca26f559b1452b32 will check for tty which is what was failing the command call test.

ashcrow commented:
@giuseppe: @jlebon noted we may want to switch to ``centos/7/atomic/alpha`` for a more stable centos test.

giuseppe commented:
@ashcrow I've just cherry-picked and pushed ad7ab9b

ashcrow commented:
Failures are related to ``test_trust.py`` which are not touched by us AFAICT.

giuseppe commented:
failures were caused by using `chdir` for running the new container.  I've changed it to use directly `Popen` and added a fixup patch to adapt the tests.  I could not leave a fixup patch for the `add support for run` as it was causing conflicts when squashed.

ashcrow commented:
We good to merge?

baude commented:
@rhatdan do you still have concerns?

rhatdan commented:
As long as this is doing atomic run and not adding atomic exec, I am fine with it.

giuseppe commented:
yes, I implemented the feature in `atomic run` and dropped completely the change that introduced `exec`

ashcrow commented:
@giuseppe Minor fix for UX https://github.com/ashcrow/atomic/commit/3430ec4dad1176600fbbae711638bf8048931535

giuseppe commented:
@ashcrow thanks!  That is much better :-)

ashcrow commented:
bot, retest this please

giuseppe commented:
tests are passing, can someone r+ it?

giuseppe commented:
hm.. I've only rebased and tests fail again.  I cherry-picked a patch from the other PR to include `make` but doesn't seem to make a difference

ashcrow commented:
bot, retest this please

ashcrow commented:
> hm.. I've only rebased and tests fail again.  I cherry-picked a patch from the other PR to include make but doesn't seem to make a difference since dnf install ... fails

cc @baude @rhatdan

rhatdan commented:
LGTM

ashcrow commented:
Odd that it installs ``PyYAML`` on CentOS and then can not import it for pylint. 🤷‍♂️

giuseppe commented:
@rh-atomic-bot delegate=ashcrow

ashcrow commented:
@rh-atomic-bot r+ 0d675c613ee5ad91efdbf225818b396d0a4e850e

ashcrow commented:
@rh-atomic-bot r+ 0d675c6

ashcrow commented:
homu couldn't find make, yet it was found previously 🤷‍♂️

ashcrow commented:
@rh-atomic-bot retry

ashcrow commented:
Received different results with no code changes. Let's try again:

@rh-atomic-bot retry

ashcrow commented:
```
### TIMED OUT AFTER 1800s
```

Known infra flake.

ashcrow commented:
@rh-atomic-bot retry

ashcrow commented:
Executing ``papr`` locally in container to see what's happening.

ashcrow commented:
This indeed a network issue. Running ``.papr`` within a container (as is done in the failing test) with this patchset applied locally works like a charm.

/cc @jlebon

ashcrow commented:
@rh-atomic-bot retry

ashcrow commented:
```
  Cannot download o/oci-systemd-hook-0.1.11-1.git1ac958a.fc26.x86_64.rpm: All mirrors were tried
```
and
```
Running test_system_containers_images.sh....    ### TIMED OUT AFTER 1800s
```

ashcrow commented:
@rh-atomic-bot retry

ashcrow commented:
This time there was just one time out ...

ashcrow commented:
@rh-atomic-bot retry

giuseppe commented:
https://github.com/projectatomic/atomic/pull/1083 solves the tests timeout issue

giuseppe commented:
@rh-atomic-bot retry

ashcrow commented:
@rh-atomic-bot retry
